## Bahaar Khalilian, Hesam Mohebi
#### Implementation of other models for comparison with KRLS

In [6]:
from sklearn.model_selection import cross_val_predict, cross_val_score, StratifiedKFold
from sklearn.metrics import mean_squared_error
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.preprocessing import StandardScaler
import warnings
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score, mean_squared_error
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import StratifiedKFold
import numpy as np
import pandas as pd

training_x_path = "trainingx.csv"
training_y_path = "trainingy.csv"

X = pd.read_csv(training_x_path, header=None)
y = pd.read_csv(training_y_path, header=None).squeeze()

#### Random Forest tree for multi-class classfication 

In [4]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score, mean_squared_error

# Initialize lists to store accuracy, confusion matrices, and classification reports for each fold
accuracy_scores = []
conf_matrices = []
class_reports = []

for train_idx, val_idx in cv.split(X, y):
    # Split data into train and validation sets for the fold
    X_train_fold, X_val_fold = X.iloc[train_idx], X.iloc[val_idx]
    y_train_fold, y_val_fold = y.iloc[train_idx], y.iloc[val_idx]
    
    # Scale the data
    X_train_scaled = scaler.fit_transform(X_train_fold)
    X_val_scaled = scaler.transform(X_val_fold)
    
    # Train the classifier on the fold
    rf_classifier.fit(X_train_scaled, y_train_fold)
    
    # Predict on train and validation sets
    y_train_pred = rf_classifier.predict(X_train_scaled)
    y_val_pred = rf_classifier.predict(X_val_scaled)
    
    # Compute MSE for the fold
    mse_train.append(mean_squared_error(y_train_fold, y_train_pred))
    mse_val.append(mean_squared_error(y_val_fold, y_val_pred))
    
    # Append accuracy, confusion matrix, and classification report
    accuracy_scores.append(accuracy_score(y_val_fold, y_val_pred))
    conf_matrices.append(confusion_matrix(y_val_fold, y_val_pred))
    class_reports.append(classification_report(
        y_val_fold, y_val_pred, output_dict=True, zero_division=0  # Suppress warning for undefined metrics
    ))

# Calculate mean MSE for train and validation
mean_mse_train = np.mean(mse_train)
mean_mse_val = np.mean(mse_val)

# Average accuracy across all folds
avg_accuracy = np.mean(accuracy_scores)

# Display metrics for the last fold as an example
conf_matrix_last_fold = pd.DataFrame(conf_matrices[-1])
class_report_last_fold = pd.DataFrame(class_reports[-1])

print("\nClassification Report (Last Fold):")
print(classification_report(y.iloc[val_idx], y_val_pred, zero_division=0))  # Suppress warning for undefined metrics
print(f"Average accuracy across all folds: {avg_accuracy:.4f}")
print(f"MSE for training set: {mean_mse_train:.4f}, MSE for validation set: {mean_mse_val:.4f}")


c:\Users\moheb\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\model_selection\_split.py:700: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(



Classification Report (Last Fold):
              precision    recall  f1-score   support

           3       0.00      0.00      0.00         5
           4       0.67      0.06      0.11        35
           5       0.70      0.71      0.70       345
           6       0.61      0.77      0.68       454
           7       0.62      0.40      0.49       169
           8       1.00      0.39      0.56        31
           9       0.00      0.00      0.00         1

    accuracy                           0.65      1040
   macro avg       0.52      0.33      0.36      1040
weighted avg       0.65      0.65      0.63      1040

Average accuracy across all folds: 0.6731
MSE for training set: 0.0000, MSE for validation set: 0.4616



#### The average accuracy accross all the folds is 67% which indicates the model performs moderately well on the data set. Yet class imbalance is profound especialy in 3 and 9 classes. 
#### The MSE on training set is 0 indicating the model has fully learn the features of training (potential overfitting), The validation MSE is 46% which suggests the model strugles with generalization speciall for underrepresented classes

## SVM implementation

In [5]:
# Define cross-validation strategy
cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

# Initialize SVM classifier
svm_classifier = SVC(class_weight='balanced', probability=True, random_state=42)

# Initialize lists to store metrics
mse_train = []
mse_val = []
accuracy_scores = []
conf_matrices = []
class_reports = []

# Scale the features
scaler = StandardScaler()

for train_idx, val_idx in cv.split(X, y):
    # Split data into train and validation sets for the fold
    X_train_fold, X_val_fold = X.iloc[train_idx], X.iloc[val_idx]
    y_train_fold, y_val_fold = y.iloc[train_idx], y.iloc[val_idx]
    
    # Scale the data
    X_train_scaled = scaler.fit_transform(X_train_fold)
    X_val_scaled = scaler.transform(X_val_fold)
    
    # Train the SVM classifier on the fold
    svm_classifier.fit(X_train_scaled, y_train_fold)
    
    # Predict on train and validation sets
    y_train_pred = svm_classifier.predict(X_train_scaled)
    y_val_pred = svm_classifier.predict(X_val_scaled)
    
    # Compute MSE for the fold
    mse_train.append(mean_squared_error(y_train_fold, y_train_pred))
    mse_val.append(mean_squared_error(y_val_fold, y_val_pred))
    
    # Append accuracy, confusion matrix, and classification report
    accuracy_scores.append(accuracy_score(y_val_fold, y_val_pred))
    conf_matrices.append(confusion_matrix(y_val_fold, y_val_pred))
    class_reports.append(classification_report(
        y_val_fold, y_val_pred, output_dict=True, zero_division=0  # Suppress warning for undefined metrics
    ))

# Calculate mean MSE for train and validation
mean_mse_train = np.mean(mse_train)
mean_mse_val = np.mean(mse_val)

# Average accuracy across all folds
avg_accuracy = np.mean(accuracy_scores)

# Display metrics for the last fold as an example
conf_matrix_last_fold = pd.DataFrame(conf_matrices[-1])
class_report_last_fold = pd.DataFrame(class_reports[-1])

print("\nClassification Report (Last Fold):")
print(classification_report(y.iloc[val_idx], y_val_pred, zero_division=0))  # Suppress warning for undefined metrics
print(f"Average accuracy across all folds: {avg_accuracy:.4f}")
print(f"MSE for training set: {mean_mse_train:.4f}, MSE for validation set: {mean_mse_val:.4f}")


c:\Users\moheb\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\model_selection\_split.py:700: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(



Classification Report (Last Fold):
              precision    recall  f1-score   support

           3       0.00      0.00      0.00         5
           4       0.14      0.57      0.23        35
           5       0.59      0.56      0.58       345
           6       0.58      0.27      0.37       454
           7       0.36      0.42      0.39       169
           8       0.12      0.61      0.21        31
           9       0.00      0.00      0.00         1

    accuracy                           0.41      1040
   macro avg       0.26      0.35      0.25      1040
weighted avg       0.52      0.41      0.43      1040

Average accuracy across all folds: 0.4232
MSE for training set: 1.1009, MSE for validation set: 1.2096


##### The average accuracy across all folds is 42.32%, which is significantly lower than the Random Forest model's performance (67.31%). This indicates that the SVM struggles with this dataset, potentially due to class imbalances or insufficient feature representation.
##### The training MSE is 1.1009, which reflects significant prediction errors on the training data. The validation MSE is 1.2096, showing that the model has difficulty generalizing to unseen data.

### Multi-layer-perseptron classifier

In [9]:
# Suppress warnings
warnings.filterwarnings("ignore", category=UserWarning)

# Define cross-validation strategy
cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

# Initialize MLP (Multi-Layer Perceptron) classifier with a more complex architecture
mlp_classifier = MLPClassifier(
    hidden_layer_sizes=(128, 64, 32),  # Three layers with decreasing neurons
    activation='relu',
    solver='lbfgs',  # Efficient solver for small to medium datasets
    max_iter=1000,  # Allow more iterations for convergence
    random_state=42,
    early_stopping=True,  # Stop training if validation score stops improving
)

# Initialize lists to store metrics
mse_train = []
mse_val = []
accuracy_scores = []
conf_matrices = []
class_reports = []

# Scale the features
scaler = StandardScaler()

for train_idx, val_idx in cv.split(X, y):
    # Split data into train and validation sets for the fold
    X_train_fold, X_val_fold = X.iloc[train_idx], X.iloc[val_idx]
    y_train_fold, y_val_fold = y.iloc[train_idx], y.iloc[val_idx]
    
    # Scale the data
    X_train_scaled = scaler.fit_transform(X_train_fold)
    X_val_scaled = scaler.transform(X_val_fold)
    
    # Train the MLP classifier on the fold
    mlp_classifier.fit(X_train_scaled, y_train_fold)
    
    # Predict on train and validation sets
    y_train_pred = mlp_classifier.predict(X_train_scaled)
    y_val_pred = mlp_classifier.predict(X_val_scaled)
    
    # Compute MSE for the fold
    mse_train.append(mean_squared_error(y_train_fold, y_train_pred))
    mse_val.append(mean_squared_error(y_val_fold, y_val_pred))
    
    # Append accuracy, confusion matrix, and classification report
    accuracy_scores.append(accuracy_score(y_val_fold, y_val_pred))
    conf_matrices.append(confusion_matrix(y_val_fold, y_val_pred))
    class_reports.append(classification_report(
        y_val_fold, y_val_pred, output_dict=True, zero_division=0  # Suppress undefined metrics
    ))

# Calculate mean MSE for train and validation
mean_mse_train = np.mean(mse_train)
mean_mse_val = np.mean(mse_val)

# Average accuracy across all folds
avg_accuracy = np.mean(accuracy_scores)

# Display metrics for the last fold as an example
conf_matrix_last_fold = pd.DataFrame(conf_matrices[-1])
class_report_last_fold = pd.DataFrame(class_reports[-1])

print("\nClassification Report (Last Fold):")
print(classification_report(y.iloc[val_idx], y_val_pred, zero_division=0))  # Suppress undefined metrics
print(f"Average accuracy across all folds: {avg_accuracy:.4f}")
print(f"MSE for training set: {mean_mse_train:.4f}, MSE for validation set: {mean_mse_val:.4f}")



Classification Report (Last Fold):
              precision    recall  f1-score   support

           3       0.17      0.20      0.18         5
           4       0.18      0.14      0.16        35
           5       0.63      0.63      0.63       345
           6       0.62      0.61      0.62       454
           7       0.50      0.55      0.53       169
           8       0.39      0.35      0.37        31
           9       0.00      0.00      0.00         1

    accuracy                           0.58      1040
   macro avg       0.36      0.36      0.35      1040
weighted avg       0.58      0.58      0.58      1040

Average accuracy across all folds: 0.5974
MSE for training set: 0.0000, MSE for validation set: 0.6999


#### The average accuracy across all folds is 59.74%, which is an improvement over the SVM model (42.32%) but still below the performance of the Random Forest model (67.31%). This suggests that the MLP captures some patterns but struggles to generalize as effectively as Random Forest.
#### The training MSE is 0.0000, indicating overfitting, as the model fits the training data perfectly. The validation MSE is 0.6999, which reflects the model's inability to generalize well, particularly on the minority classes.